### DISCUSSION TITLE
Have quarto find the jupyter kernel within Visual Studio Code

### DISCUSSION THREAD

**Discussion by JaumeAmoresDS** [quarto-dev/quarto-cli#10026] [Q&A]
### Description

I am trying to follow this [vscode tutorial](https://quarto.org/docs/get-started/hello/vscode.html)  in the quarto.org [site](https://quarto.org). 

The tutorial uses a document `hello.qmd` whose content can be copied from the same tutorial. We need to edit the `hello.qmd` to indicate the jupyter kernel we will be using. Let us assume it is `my_kernel`, we change the header to be: 

```
---
title: "Quarto Basics"
format:
  html:
    code-fold: true
jupyter: my_kernel
---
```

Now, the trick is to have quarto find the kernel `my_kernel`. If I just try to preview the document (with the quarto extension installed as indicated in the tutorial), I get the following type of error:

```
ERROR: Jupyter kernel my_kernel not found. Known kernels: python3. Run 'quarto check jupyter' with your python environment activated to check python version used. 
```

Then if I run `quarto check jupyter` I get:
```
Jupyter is not available in this Python installation.
```

The issue seems that quarto preview is run within the `base` conda environment. In the mentioned tutorial, we are instructed to install jupyter in this base environment. However, I don't want to install anything in that base environment (I never do it). Does anyone know of an easy way to indicate the conda environment to be used, the one associated with the kernel for running the code in the qmd file? 

I know the question might be more related with VS Code settings, but any help is appreciated. Thanks!

[**Note**: In my case, **I am running VS Code from within WSL** (by using the remote VSCode connection `WSL:Ubuntu`)].

**Comment by mcanouil** [comment]
Did you read the following: https://quarto.org/docs/computations/python.html#kernel-selection

In particular the "important note"?

Also, If you run `jupyter kernelspeclist` do you see your kernel? If not then Quarto can't either.

**Reply by JaumeAmoresDS** [reply]
Thanks @mcanouil. As I mention in my question, I am already indicating the right jupyter kernel in the qmd document. I also mention that jupyter is not accessible from the Quarto extension of VSC, the reason being that it is not installed in the default base environment. Finallly, I mention that the solution given in the tutorial is to install jupyter, but I don't want to install libraries in my base environment.

The solution to my question is more on the lines of using the right set up for VSC, nothing to do with qmd or environment. I know some possibilites but none of them are ideal, so that's why I am asking. Since this is about having quarto running on VSC, I figured it would be good to let the community know about this type of question.

**Reply by JaumeAmoresDS** [reply]
It is not necessary to run `jupyter kernelspeclist`. Just running `jupyter check` as suggested in the error message (also included in my question) is enough.

**Reply by mcanouil** [reply]
So from what you are saying, if you run Quarto in the command line, it works?

**Reply by JaumeAmoresDS** [reply]
Yes, if I run it after activating an environment which has jupyter installed in it. In other words, I can work with quarto with no issues. The difficulty comes from having it integrated within my VSC environment, as done in the tutorial whose link I included.

**Reply by mcanouil** [reply]
So this is unrelated to Quarto.
You need to check the Python path set in your VSCode settings.
And you still need to activate yourself the environment.
VSCode only activates virtual environment set inside the workspace, any environment outside the workspace has to be activated by the user. You should check VSCode documentation for this.

**Reply by JaumeAmoresDS** [reply]
I would like to disagree (respectfully :-)) on both statements. As you can see in the extensive and very nice documentation from the quarto site, this type of setup questions, related to VS Code, RStudio, nvim, and others, are covered in order to help users setup their environment to work have quarto integrated in it seamlessly. In fact, in the tutorial I refer to they do discuss the issue of setting up VS Code to work with jupyter (i.e., "unrelated" to quarto). 

Related with the second statement, it is not correct that "VSCode only activates virtual environment set inside the workspace, any environment outside the workspace has to be activated by the user" as you say. In fact, one possible solution (not ideal, but solution nonetheless) to avoid installing jupyter in the base conda environment is as follows:

In order to have VS Code use the conda environment where jupyter is installed (other than the default base environment) we can  make VS Code automatically pick up our preferred environment everytime we open up a new terminal, as explained [here](https://github.com/microsoft/vscode-python/wiki/Setting-descriptions#pythondefaultinterpreterpath). This is done as follows:

In the Command Palette (Ctrl+Shift+P), I select "Preferences: Open Workspace Settings", then go to the Extensions section on the left, select Python, and then edit the "Default Interpreter" path to point to the python executable used by my environment. In Linux, we can find the later by activating the environment and then running `which python`:

```bash
conda activate my_environment
which python
```

which, in my case, indicates something like:

```
/home/my_user/miniconda3/envs/my_environment/bin/python
```

**Reply by mcanouil** [reply]
Quarto tutorials are a great starting point for the basics. While they don't cover every setup and configuration of third-party tools (to avoid a hefty maintenance load), they do provide a solid foundation. For instance, not all IDEs are detailed, but you'll get the gist of it.

Regarding your VSCode setup, it's best to refer to the Python path in your settings, as mentioned ("You need to check the Python path set in your VSCode settings." and "You should check VSCode documentation for this."). The VSCode and Python extension documentation will guide you through changing the Python path.

Remember, setting up your environment is a separate process from Quarto CLI's detection capabilities. Quarto CLI will recognise any active environment, but it's not responsible for activating or setting up your environment—that part's on you. Hope this helps! 😊

Now, if you think some things are missing you have two options:
- Open "documentation" issue using the issue template. In this issue, you'll describe what is missing and what you would add.
- Open a Pull Request on the documentation with your improvement proposal.

**Comment by JaumeAmoresDS** [comment]
To see what I am referring to, could you please just have a look at this link from the quarto docs:

https://quarto.org/docs/get-started/hello/vscode.html

which is the one I am referring to all along, and the one we're discussing about. In particular, please read the following in that quarto page:

![image](https://github.com/quarto-dev/quarto-cli/assets/124528012/9a3ce970-7544-4986-9275-a31c428a50bc)

Now, you could say that the document is discussing about things that are not strictly related with Quarto, but about installing jupyter and so on. And I can see your point. But at the  same time I would argue that the objective of this documentation is actually to help users set up VS Code, including how  to install jupyter so that VS Code picks it up, which in turn is a necessary prerequisite to be able to have quarto render qmd documents that have python code. 

If you agree with that point of view (even if it is not yours, which I totally respect), then you can see that it is of interest to this community to discuss how to do the same thing but not  within the default base conda environment, since this is an anti-pattern.  You might agree or not with that, but the question and possible answers are still of interest to many people who want to use Quarto integrated with VS Code without polluting the base environment. Note that this is rarely needed when using VS Code outside of the current case.

**Reply by mcanouil** [reply]
A default setup (even using Conda or any virtual environment) just works following the tutorial.

 It appears that your requirements extend beyond the standard default configuration.
The primary objective of a “get started” tutorial is to facilitate a swift and straightforward initiation into the most commonly employed setup, rather than to deliver an exhaustive guide on intricate setup procedures.
For those seeking to delve into more sophisticated configurations, our documentation offers guidance, although for comprehensive understanding, consulting the original documentation is advisable.
Have you had the opportunity to consult the following resources?

- https://quarto.org/docs/projects/virtual-environments.html
- https://quarto.org/docs/computations/python.html

Could you please specify what you believe is lacking in the screenshot you have provided? Your feedback is invaluable to us as we don't believe the documentation is perfect.

**Reply by JaumeAmoresDS** [reply]
The point here is that many users, well versed in python, do not consider what you think "standard" as the actual standard, but as an anti-pattern, and therefore it is of interest to follow what is usually considered best practices in terms of python virtual environments. Again, this might not be your particular point of view, but that doesn't invalidate the importance of this question and the seek for an answer from other quarto users who might have a good solution. If that's not your case, it is completely fine, but could I respectfully ask to let other people answer this question if they find it relevant?

If you could on your side just google about best practices in terms of python virtual environments you might see what I refer to, but I would honestly rather not spend too much time on whether or not this follows your point of view?

Regarding your questions, yes, i did look at all these sections of the tutorial.
